File: core/src/main/java/edu/knowitall/taggers/tag/KeywordTagger.java
package edu.knowitall.taggers.tag;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

import com.google.common.base.Function;
import com.google.common.base.Joiner;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;

import edu.knowitall.collection.immutable.Interval;
import edu.knowitall.taggers.SentenceFunctions;
import edu.knowitall.taggers.StringFunctions;
import edu.knowitall.tool.typer.Type;
import edu.knowitall.tool.chunk.ChunkedToken;
import edu.knowitall.tool.stem.Lemmatized;

/***
 * Search for exact keyword matches, case-sensitive, and tag the match.
 * @author schmmd
 *
 */
public class KeywordTagger extends JavaTagger {
    public ImmutableList<String[]> keywords;

    public KeywordTagger(String name, Collection<String> keywords) {
        super(name, null);

        setKeywords(new TreeSet<String>(keywords));
    }

    /**
     * Constructor used by reflection.
     * @param name name of the tagger
     * @param args arguments to the tagger
     */
    public KeywordTagger(String name, scala.collection.Seq<String> args) {
        this(name, scala.collection.JavaConversions.asJavaList(args));
    }

    protected void setKeywords(Set<String> keywords) {
        this.keywords = ImmutableList.copyOf(
                Iterables.transform(keywords, StringFunctions.split("\\s+")));
    }

    protected void setKeywords(Iterable<String> keywords) {
        this.keywords = ImmutableList.copyOf(
                Iterables.transform(keywords, StringFunctions.split("\\s+")));
    }

    protected List<String> getKeywords() {
        return Lists.transform(this.keywords, new Function<String[], String>() {
            @Override
            public String apply(String[] parts) {
                return Joiner.on(" ").join(parts);
            }
        });
    }

    @Override
    public boolean equals(Object that) {
        if (that == null) return false;
        if (this == that) return true;
        if (this.getClass() != that.getClass()) return false;

        KeywordTagger kt = (KeywordTagger)that;
        if (!super.equals(that)) {
            return false;
        }

        if (this.keywords.size() != kt.keywords.size()) {
            return false;
        }

        if (this.keywords.size() != kt.keywords.size()) {
            return false;
        }

        Iterator<String[]> i1 = this.keywords.iterator();
        Iterator<String[]> i2 = kt.keywords.iterator();

        while (i1.hasNext() && i2.hasNext()) {
            if (!Arrays.equals(i1.next(), i2.next())) {
                return false;
            }
        }

        return true;
    }

    @Override
    public List<Type> findTagsJava(final List<Lemmatized<ChunkedToken>> sentence) {
        return this.findTagsJava(sentence, SentenceFunctions.strings(sentence));
    }

    protected List<Type> findTagsJava(final List<Lemmatized<ChunkedToken>> sentence,
            List<String> tokens) {
        List<Type> tags = new ArrayList<Type>();
        for (String[] keyword : keywords) {
            this.findTagsJava(tags, sentence, tokens, keyword);
        }
        return tags;
    }

    protected void findTagsJava(final List<Type> tags, final List<Lemmatized<ChunkedToken>>  sentence,
            List<String> tokens, final String[] keyword) {

        for (int i = 0; i <= tokens.size() - keyword.length; i++) {
            if (match(tokens, keyword, i)) {
                Interval range = Interval.open(i, i + keyword.length);
                tags.add(this.createType(sentence, range));
            }
        }
    }

    protected boolean match(List<String> source, String[] target, int sourceIndex) {
        int targetIndex = 0;
        for (String targetWord : target) {
            String sourceWord = source.get(sourceIndex + targetIndex);

            if (!sourceWord.equals(targetWord)) {
                return false;
            }

            targetIndex++;
        }

        return true;
    }
}

File: core/src/main/scala/org/allenai/taggers/tag/KeywordTagger.scala
package org.allenai.taggers.tag

import com.google.common.base.Function
import com.google.common.base.Joiner
import com.google.common.collect.ImmutableList
import com.google.common.collect.Iterables
import com.google.common.collect.Lists
import org.allenai.common.immutable.Interval
import org.allenai.nlpstack.core.ChunkedToken
import org.allenai.nlpstack.core.Lemmatized
import org.allenai.nlpstack.core.typer.Type
import org.allenai.taggers.SentenceFunctions
import org.allenai.taggers.StringFunctions
import org.allenai.nlpstack.core.repr._

class KeywordTagger(val name: String, val wholeKeywords: Seq[String])
extends Tagger[Tagger.Sentence with Chunks] {
  override def source = null

  // Keywords are stored as a map for more efficient matching in a sentence.
  // The key of the map is the first token of the keyword.
  // The value of the map is a seq of keywords by token.
  private val keywordMap: Map[String, Seq[List[String]]] = {
    val keywords = wholeKeywords map (_.split("\\s+").toList)
    keywords.groupBy(_.take(1)).map { case (k, v) => k.head -> v }.toMap
  }

  override def equals(that: Any) = super.equals(that) && (that match {
    case that: KeywordTagger => wholeKeywords == that.wholeKeywords
    case _ => false
  })
  override def canEqual(that: Any) = that.isInstanceOf[KeywordTagger]

  override def tag(sentence: TheSentence, types: Seq[Type]): Seq[Type] = {
    keywordTag(sentence.strings, types)
  }

  /** This method exists so the sentence can be transformed before matching against the keywords. 
    * For example, the sentence might be lower cased so a case insensitive match can be done
    * against the keywords (which would already have been lower cased as well).
    */
  protected def keywordTag(sentence: Seq[String], types: Seq[Type]): Seq[Type] = {
    /** Try and match a keyword at a particular index. */
    def isMatch(target: Seq[String], sourceIndex: Int): Boolean = {
      var targetIndex = 0

      for (targetWord <- target) {
        if (sourceIndex + targetIndex >= sentence.size) {
          return false
        }

        val sourceWord = sentence(sourceIndex + targetIndex)

        if (!sourceWord.equals(targetWord)) {
          return false
        }

        targetIndex += 1
      }

      return true
    }

    /** Try and match a keyword at a particular index. 
      * If there is a match, create and return a Type.
      */
    def matchKeyword(keyword: Seq[String], i: Int): Option[Type] = {
      if (isMatch(keyword, i)) {
        val interval = Interval.open(i, i + keyword.length)
        val text = sentence.slice(interval.start, interval.end).mkString(" ")
        Some(Type(this.name, this.source, interval, text))
      }
      else {
        None
      }
    }

    for {
      // Iterate over the sentence tokens.
      i <- 0 until sentence.size
      // Iterate over any keyword where the first token matches the current token in the sentence.
      keyword <- keywordMap.get(sentence(i)).getOrElse(Seq.empty)
      // Check to see if there is a match with all the tokens in the keyword.
      typ <- matchKeyword(keyword, i)
    } yield typ
  }
}
